### ISSUE TITLE
Enhancement: detect safe navigation in `Style/ReturnNilInPredicateMethodDefinition`

### ISSUE THREAD

**Issue by vlad-pisanov** [rubocop/rubocop#13114]
Enhancement idea: if the return value of a predicate contains safe navigation `&`, `Style/ReturnNilInPredicateMethodDefinition` should flag it as an offense:

```ruby
def allowed?
  user&.admin?   # ⚠️ predicate might return nil
end
```

**Comment by dvandersluis** [comment]
I started looking into this but I'm not sure what the proper solution here would be. I was thinking the "good" case might be:

```ruby
def allowed
  return false unless user
  user.admin?
end
```

However, this causes quite a few offenses in the various rubocop repos that I don't know if they are improvements. For example https://github.com/rubocop/rubocop-ast/blob/737c62441162043a551fc624cf7f6caccbf8e319/lib/rubocop/ast/node/block_node.rb#L89-L101

We could do this as a configuration option, but I'm not sure of the value.  I have code written to do the above, and if it is desired I can open a PR. What do you think @rubocop/rubocop-core?

In the meantime, I found some other situations where `Style/ReturnNilInPredicateMethodDefinition` could be useful that I'm going to open a separate PR for.

**Comment by Earlopain** [comment]
The good case would actually be

```rb
def allowed`
  return false unless user

  user.admin?
end
```

which I agree doesn't have the same ring to it as just `user.&admin?`. `Layout/EmptyLineAfterGuardClause` is enabled by default and also one of the less popular cops I'd say (together with `Style/GuardClause`).

It's easy to correct implicit/explicit returns without safe navigation, the code is already there and you just insert `true`/`false`. `nil` is fine as a substitue for `false`.

As an alternative, the correction could be something like `user&.admin? || false` but it feels redundant (because it is :upside_down_face:).

**Comment by marcandre** [comment]
> I started looking into this but I'm not sure what the proper solution here would be. I was thinking the "good" case might be:
> 
> ```ruby
> def allowed
>   return false unless user
>   user.admin?
> end
> ```
> 
> However, this causes quite a few offenses in the various rubocop repos that I don't know if they are improvements. For example https://github.com/rubocop/rubocop-ast/blob/737c62441162043a551fc624cf7f6caccbf8e319/lib/rubocop/ast/node/block_node.rb#L89-L101

These are good examples 🚀 Either:
1) The `?` is not helpful
2) The doc should be changed
3) The code should be changed

I believe in both cases this is 1)

**Comment by dvandersluis** [comment]
@earlopain yes I mistyped, it should be `def allowed?` not `def allowed` (ie it was intended to remain a predicate).

**Comment by dvandersluis** [comment]
@marcandre thanks - there are a lot of examples in rubocop/rubocop as well. For instance https://github.com/rubocop/rubocop/blob/2c0f759c96c11a8da7811eec9904436c46632c02/lib/rubocop/cop/lint/symbol_conversion.rb#L139-L141

**Comment by Earlopain** [comment]
> yes I mistyped, it should be def allowed? not def allowed (ie it was intended to remain a predicate).

Yes, of course. I was more meaning that the correction for the majority of user would also imply a newline because of `Layout/EmptyLineAfterGuardClause`

**Comment by dvandersluis** [comment]
> I was more meaning that the correction for the majority of user would also imply a newline because of `Layout/EmptyLineAfterGuardClause`

Yes of course. I was just demonstrating the change this cop specifically would make but you're right that the change would take 1 line methods and make them 3 lines in many cases.

**Comment by vlad-pisanov** [comment]
What about simply recommending
```ruby
!user.nil? && user.admin?
```
instead of a guard clause? This pattern is simpler to autocorrect, and is composable

**Comment by Earlopain** [comment]
To me, that code just is uglier than before. There's nothing really wrong with returning `nil` in a predicate. I feel like this is something better addressed by tools like sorbet or steep. They will do a better job of ensuring consistent return types anyways.

**Comment by bquorning** [comment]
Using a null object, you could still write
```rb
def allowed?
  user.admin?
end
```
If not, perhaps `!!` is the cleaner way to write it? (I know we disallow `!!` for some reason, but I never understood why.)
```rb
def allowed?
  !!user&.admin?
end
```

**Comment by dvandersluis** [comment]
If `Style/DoubleNegation` is enabled with `EnforcedStyle: forbidden`, full autocorrection will result in

```ruby
def allowed?
  !user&.admin?.nil?
end
```
